File: cowboy/src/cards/card_handler.erl
-module(card_handler).
-behaviour(cowboy_rest).

-export([init/2, allowed_methods/2, content_types_provided/2, content_types_accepted/2, resource_exists/2, handle_get/2, handle_post/2, allow_missing_post/2, handle_put/2, handle_patch/2, handle_ban/2, handle_unban/2, handle_restrict/2, handle_unrestrict/2, handle_calculate_value/2, handle_apply_rarity_bonus/2, handle_is_legal_in_format/2]).

-include("records.hrl").

init(Req, State) ->
    {cowboy_rest, Req, State}.

allowed_methods(Req, State) ->
    {[<<"POST">>, <<"GET">>, <<"PUT">>, <<"PATCH">>], Req, State}.

content_types_provided(Req, State) ->
    {[{<<"application/json">>, handle_get}], Req, State}.

content_types_accepted(Req, State) ->
    Method = cowboy_req:method(Req),
    Handler = case Method of
        <<"PUT">>   -> handle_put;
        <<"PATCH">> -> handle_patch;
        _           -> handle_post
    end,
    {[{<<"application/json">>, Handler}], Req, State}.

resource_exists(Req, State) ->
    case cowboy_req:binding(id, Req) of
        undefined -> {true, Req, State};
        Id ->
            case card_store:find(binary_to_integer(Id)) of
                {ok, Record} -> {true, Req, Record};
                not_found    -> {false, Req, State}
            end
    end.

allow_missing_post(Req, State) -> {false, Req, State}.

handle_get(Req, State) ->
    case cowboy_req:binding(id, Req) of
        undefined ->
            Qs = cowboy_req:parse_qs(Req),
            Q  = proplists:get_value(<<"q">>, Qs, <<"">>),
            All = card_store:all(),
            Filtered = [R || R <- All, Q =:= <<"">> orelse binary:match(maps:get(name, R, <<"">>), Q) =/= nomatch],
            Body = jsone:encode(Filtered),
            {Body, Req, State};
        _Id ->
            Body = jsone:encode(State),
            {Body, Req, State}
    end.

handle_post(Req0, State) ->
    {ok, Body, Req1} = cowboy_req:read_body(Req0),
    Params = jsone:decode(Body, [{object_format, map}]),
    case validate_card_rules(Params) of {error, Errs} -> reply_422(Req1, Errs, State); ok ->
    case validate_card_implies(Params) of {error, Errs} -> reply_422(Req1, Errs, State); ok ->
    Id     = card_store:next_id(),
    Record = params_to_record(Id, Params),
    Record1 = validate_legality_hook(Record),
    ok     = card_store:insert(Record1),
    Resp   = jsone:encode(record_to_map(Record)),
    Req2   = cowboy_req:reply(201, #{<<"content-type">> => <<"application/json">>}, Resp, Req1),
    {stop, Req2, State}
    end end
.

handle_put(Req0, State) ->
    case is_map(State) andalso maps:is_key(<<"id">>, State) of
        false -> {stop, cowboy_req:reply(404, #{}, <<>>, Req0), State};
        true  ->
    Id = maps:get(<<"id">>, State),
    {ok, ExistingRecord} = card_store:find_record(Id),
    {ok, Body, Req1} = cowboy_req:read_body(Req0),
    Params = jsone:decode(Body, [{object_format, map}]),
    Updated = merge_record(ExistingRecord, Params),
    Updated1 = validate_legality_hook(Updated),
    ok = card_store:update(Updated1),
    Resp = jsone:encode(record_to_map(Updated)),
    Req2 = cowboy_req:reply(200, #{<<"content-type">> => <<"application/json">>}, Resp, Req1),
    {stop, Req2, Updated}
    end.

handle_patch(Req0, State) -> handle_put(Req0, State).

params_to_record(Id, Params) ->
    #card{
        id         = Id,
        name       = maps:get(<<"name">>, Params, undefined),
        card_type  = maps:get(<<"card_type">>, Params, undefined),
        rarity     = maps:get(<<"rarity">>, Params, undefined),
        mana_cost  = maps:get(<<"mana_cost">>, Params, undefined),
        mana_colors = maps:get(<<"mana_colors">>, Params, undefined),
        attack     = maps:get(<<"attack">>, Params, undefined),
        defense    = maps:get(<<"defense">>, Params, undefined),
        loyalty    = maps:get(<<"loyalty">>, Params, undefined),
        description = maps:get(<<"description">>, Params, undefined),
        flavor_text = maps:get(<<"flavor_text">>, Params, undefined),
        image_url  = maps:get(<<"image_url">>, Params, undefined),
        artist_name = maps:get(<<"artist_name">>, Params, undefined),
        legal_formats = maps:get(<<"legal_formats">>, Params, undefined),
        is_banned  = maps:get(<<"is_banned">>, Params, undefined),
        is_restricted = maps:get(<<"is_restricted">>, Params, undefined),
        power_level = maps:get(<<"power_level">>, Params, undefined),
        set_id     = maps:get(<<"set_id">>, Params, undefined),
        created_at = iso_now(),
        updated_at = iso_now()
    }.

merge_record(Record, Params) ->
    #card{
        id         = Record#card.id,
        name       = maps:get(<<"name">>, Params, Record#card.name),
        card_type  = maps:get(<<"card_type">>, Params, Record#card.card_type),
        rarity     = maps:get(<<"rarity">>, Params, Record#card.rarity),
        mana_cost  = maps:get(<<"mana_cost">>, Params, Record#card.mana_cost),
        mana_colors = maps:get(<<"mana_colors">>, Params, Record#card.mana_colors),
        attack     = maps:get(<<"attack">>, Params, Record#card.attack),
        defense    = maps:get(<<"defense">>, Params, Record#card.defense),
        loyalty    = maps:get(<<"loyalty">>, Params, Record#card.loyalty),
        description = maps:get(<<"description">>, Params, Record#card.description),
        flavor_text = maps:get(<<"flavor_text">>, Params, Record#card.flavor_text),
        image_url  = maps:get(<<"image_url">>, Params, Record#card.image_url),
        artist_name = maps:get(<<"artist_name">>, Params, Record#card.artist_name),
        legal_formats = maps:get(<<"legal_formats">>, Params, Record#card.legal_formats),
        is_banned  = maps:get(<<"is_banned">>, Params, Record#card.is_banned),
        is_restricted = maps:get(<<"is_restricted">>, Params, Record#card.is_restricted),
        power_level = maps:get(<<"power_level">>, Params, Record#card.power_level),
        set_id     = maps:get(<<"set_id">>, Params, Record#card.set_id),
        created_at = Record#card.created_at,
        updated_at = iso_now()
    }.

record_to_map(#card{id = Id, name = Name, card_type = CardType, rarity = Rarity, mana_cost = ManaCost, mana_colors = ManaColors, attack = Attack, defense = Defense, loyalty = Loyalty, description = Description, flavor_text = FlavorText, image_url = ImageUrl, artist_name = ArtistName, legal_formats = LegalFormats, is_banned = IsBanned, is_restricted = IsRestricted, power_level = PowerLevel, set_id = SetId, created_at = CreatedAt, updated_at = UpdatedAt}) ->
    #{
        <<"id">> => Id,
        <<"name">> => Name,
        <<"card_type">> => CardType,
        <<"rarity">> => Rarity,
        <<"mana_cost">> => ManaCost,
        <<"mana_colors">> => ManaColors,
        <<"attack">> => Attack,
        <<"defense">> => Defense,
        <<"loyalty">> => Loyalty,
        <<"description">> => Description,
        <<"flavor_text">> => FlavorText,
        <<"image_url">> => ImageUrl,
        <<"artist_name">> => ArtistName,
        <<"legal_formats">> => LegalFormats,
        <<"is_banned">> => IsBanned,
        <<"is_restricted">> => IsRestricted,
        <<"power_level">> => PowerLevel,
        <<"set_id">> => SetId,
        <<"created_at">> => CreatedAt,
        <<"updated_at">> => UpdatedAt
    }.

iso_now() ->
    {{Y,Mo,D},{H,Mi,S}} = calendar:universal_time(),
    iolist_to_binary(io_lib:format("~4..0B-~2..0B-~2..0BT~2..0B:~2..0B:~2..0BZ", [Y,Mo,D,H,Mi,S])).

reply_422(Req, Errors, State) ->
    Body = jsone:encode(#{<<"errors">> => Errors}),
    Req2 = cowboy_req:reply(422, #{<<"content-type">> => <<"application/json">>}, Body, Req),
    {stop, Req2, State}.

%% ── Lifecycle hooks ──────────────────────────────────────────────────
validate_legality_hook(Record) ->
    %% TODO: implement validate_legality
    Record.

to_number(V) when is_integer(V) -> float(V);
to_number(V) when is_float(V)   -> V;
to_number(V) when is_binary(V)  ->
    try binary_to_float(V) catch _:_ -> try float(binary_to_integer(V)) catch _:_ -> 0.0 end end;
to_number(_)                    -> 0.0.

validate_card_rules(M) ->
    Checks = [
        fun() -> case (to_number(maps:get(<<"mana_cost">>, M, undefined)) >= 0 andalso to_number(maps:get(<<"mana_cost">>, M, undefined)) =< 20) of false -> {true, <<"mana_cost must be between 0 and 20">>}; _ -> false end end,
        fun() -> case (to_number(maps:get(<<"power_level">>, M, undefined)) >= 1 andalso to_number(maps:get(<<"power_level">>, M, undefined)) =< 10) of false -> {true, <<"power_level must be between 1 and 10">>}; _ -> false end end,
        fun() -> case (not (((maps:get(<<"is_banned">>, M, undefined) =:= true) andalso (maps:get(<<"is_restricted">>, M, undefined) =:= true)))) of false -> {true, <<"Card cannot be both banned and restricted at the same time">>}; _ -> false end end
    ],
    Errors = lists:filtermap(fun(F) -> F() end, Checks),
    case Errors of
        [] -> ok;
        _  -> {error, Errors}
    end.

validate_card_implies(M) ->
    Checks = [
        fun() -> case ((maps:get(<<"card_type">>, M, undefined) =:= <<"Creature">>)) andalso not (((maps:get(<<"attack">>, M, undefined) =/= undefined andalso maps:get(<<"attack">>, M, undefined) =/= null) andalso (maps:get(<<"defense">>, M, undefined) =/= undefined andalso maps:get(<<"defense">>, M, undefined) =/= null))) of true -> {true, <<"Creature card must have attack and defense">>}; _ -> false end end,
        fun() -> case ((maps:get(<<"card_type">>, M, undefined) =:= <<"Planeswalker">>)) andalso not ((maps:get(<<"loyalty">>, M, undefined) =/= undefined andalso maps:get(<<"loyalty">>, M, undefined) =/= null)) of true -> {true, <<"Planeswalker card must have loyalty">>}; _ -> false end end,
        fun() -> case ((maps:get(<<"card_type">>, M, undefined) =:= <<"Land">>)) andalso not ((maps:get(<<"mana_cost">>, M, undefined) =:= 0)) of true -> {true, <<"Land card must have zero mana cost">>}; _ -> false end end,
        fun() -> case ((maps:get(<<"card_type">>, M, undefined) =/= <<"Planeswalker">>)) andalso not ((maps:get(<<"loyalty">>, M, undefined) =:= undefined orelse maps:get(<<"loyalty">>, M, undefined) =:= null)) of true -> {true, <<"Only Planeswalker cards can have loyalty">>}; _ -> false end end,
        fun() -> case ((maps:get(<<"is_banned">>, M, undefined) =:= true)) andalso not ((maps:get(<<"legal_formats">>, M, undefined) =:= <<"message">>)) of true -> {true, <<"banned_card_not_in_legal_formats">>}; _ -> false end end
    ],
    Errors = lists:filtermap(fun(F) -> F() end, Checks),
    case Errors of
        [] -> ok;
        _  -> {error, Errors}
    end.

%% ── Behavior endpoints ──────────────────────────────────────────────
handle_ban(Req, State) ->
    _ = ban_behavior(State),
    {true, cowboy_req:reply(204, #{}, <<>>, Req), State}.

ban_behavior(_Record) ->
    %% TODO: implement ban
    ok.

handle_unban(Req, State) ->
    _ = unban_behavior(State),
    {true, cowboy_req:reply(204, #{}, <<>>, Req), State}.

unban_behavior(_Record) ->
    %% TODO: implement unban
    ok.

handle_restrict(Req, State) ->
    _ = restrict_behavior(State),
    {true, cowboy_req:reply(204, #{}, <<>>, Req), State}.

restrict_behavior(_Record) ->
    %% TODO: implement restrict
    ok.

handle_unrestrict(Req, State) ->
    _ = unrestrict_behavior(State),
    {true, cowboy_req:reply(204, #{}, <<>>, Req), State}.

unrestrict_behavior(_Record) ->
    %% TODO: implement unrestrict
    ok.

handle_calculate_value(Req, State) ->
    Result = calculate_value_behavior(State),
    Resp = jsone:encode(#{<<"result">> => Result}),
    {true, cowboy_req:reply(200, #{<<"content-type">> => <<"application/json">>}, Resp, Req), State}.

calculate_value_behavior(_Record) ->
    %% TODO: implement calculate_value
    null.

handle_apply_rarity_bonus(Req0, State) ->
    {ok, Body, Req1} = cowboy_req:read_body(Req0),
    Params = jsone:decode(Body, [{object_format, map}]),
    Result = apply_rarity_bonus_behavior(State, Params),
    Resp = jsone:encode(#{<<"result">> => Result}),
    {true, cowboy_req:reply(200, #{<<"content-type">> => <<"application/json">>}, Resp, Req1), State}.

apply_rarity_bonus_behavior(_Record, _Params) ->
    %% TODO: implement apply_rarity_bonus(multiplier)
    null.

handle_is_legal_in_format(Req0, State) ->
    {ok, Body, Req1} = cowboy_req:read_body(Req0),
    Params = jsone:decode(Body, [{object_format, map}]),
    Result = is_legal_in_format_behavior(State, Params),
    Resp = jsone:encode(#{<<"result">> => Result}),
    {true, cowboy_req:reply(200, #{<<"content-type">> => <<"application/json">>}, Resp, Req1), State}.

is_legal_in_format_behavior(_Record, _Params) ->
    %% TODO: implement is_legal_in_format(format)
    null.

File: dream/lib/cards_project/cards/card_handler.ml
(* Dream handlers for Card *)
open Lwt.Syntax

(* ── Lifecycle hooks ─────────────────────────────────────────────────── *)
let hook_validate_legality () =
  (* TODO: implement validate_legality *)
  ()

let hook_validate_not_in_use () =
  (* TODO: implement validate_not_in_use *)
  ()

(* JSON field helpers *)
let json_string_opt j key =
  match Yojson.Safe.Util.member key j with
  | `String s -> Some s
  | `Null | `Assoc [] -> None
  | _ -> None

let json_float_opt j key =
  match Yojson.Safe.Util.member key j with
  | `Float f -> Some f
  | `Int i -> Some (float_of_int i)
  | `String s -> (try Some (float_of_string s) with _ -> None)
  | _ -> None

let json_present j key =
  match Yojson.Safe.Util.member key j with
  | `Null | `Assoc [] -> false
  | _ -> true

let json_bool_opt j key =
  match Yojson.Safe.Util.member key j with
  | `Bool b -> Some b
  | _ -> None

let validate_card (j : Yojson.Safe.t) : (unit, string list) result =
  let errors = ref [] in
  if not ((match (json_float_opt j "mana_cost") with Some v -> v >= 0. && v <= 20. | None -> true)) then errors := "mana_cost must be between 0 and 20" :: !errors;
  if not ((match (json_float_opt j "power_level") with Some v -> v >= 1. && v <= 10. | None -> true)) then errors := "power_level must be between 1 and 10" :: !errors;
  if not (not (((json_bool_opt j "is_banned") = Some true && (json_bool_opt j "is_restricted") = Some true))) then errors := "Card cannot be both banned and restricted at the same time" :: !errors;
  if not ((not ((json_string_opt j "card_type") = Some "Creature") || ((json_present j "attack") && (json_present j "defense")))) then errors := "Creature card must have attack and defense" :: !errors;
  if not ((not ((json_string_opt j "card_type") = Some "Planeswalker") || ((json_present j "loyalty")))) then errors := "Planeswalker card must have loyalty" :: !errors;
  if not ((not ((json_string_opt j "card_type") = Some "Land") || ((json_float_opt j "mana_cost") = Some 0.))) then errors := "Land card must have zero mana cost" :: !errors;
  if not ((not ((json_string_opt j "card_type") <> Some "Planeswalker") || ((not (json_present j "loyalty"))))) then errors := "Only Planeswalker cards can have loyalty" :: !errors;
  if not ((not ((json_bool_opt j "is_banned") = Some true) || ((json_string_opt j "legal_formats") = Some "message"))) then errors := "banned_card_not_in_legal_formats" :: !errors;
  if !errors = [] then Ok () else Error (List.rev !errors)

let extract_insert_params (j : Yojson.Safe.t) =
  let open Yojson.Safe.Util in
  let name = match member "name" j with `String s -> s | _ -> "" in
  let card_type = match member "card_type" j with `String s -> s | _ -> "" in
  let rarity = match member "rarity" j with `String s -> s | _ -> "" in
  let mana_cost = match member "mana_cost" j with `Int i -> i | _ -> 0 in
  let mana_colors = match member "mana_colors" j with `String s -> s | _ -> "" in
  let attack = match member "attack" j with `Int i -> Some i | _ -> None in
  let defense = match member "defense" j with `Int i -> Some i | _ -> None in
  let loyalty = match member "loyalty" j with `Int i -> Some i | _ -> None in
  let description = match member "description" j with `String s -> s | _ -> "" in
  let flavor_text = match member "flavor_text" j with `String s -> Some s | _ -> None in
  let image_url = match member "image_url" j with `String s -> Some s | _ -> None in
  let artist_name = match member "artist_name" j with `String s -> Some s | _ -> None in
  let legal_formats = match member "legal_formats" j with `String s -> s | _ -> "" in
  let is_banned = match member "is_banned" j with `Bool b -> b | _ -> false in
  let is_restricted = match member "is_restricted" j with `Bool b -> b | _ -> false in
  let power_level = match member "power_level" j with `Int i -> i | _ -> 0 in
  let set_id = match member "set_id" j with `Int i -> i | _ -> 0 in
  (((name, card_type, rarity, mana_cost), (mana_colors, attack, defense, loyalty), (description, flavor_text, image_url, artist_name), (legal_formats, is_banned, is_restricted, power_level)), set_id)

let handler_card (db : (module Caqti_lwt.CONNECTION)) req =
  let respond_json status body =
    Dream.respond ~status:(Dream.int_to_status status)
      ~headers:[("Content-Type", "application/json")]
      (Yojson.Safe.to_string body)
  in
  let module Db = (val db) in
  let target_path = match String.index_opt (Dream.target req) '?' with
    | Some i -> String.sub (Dream.target req) 0 i
    | None -> Dream.target req
  in
  let path_segments = target_path |> String.split_on_char '/' |> List.filter (fun s -> s <> "") in
  match Dream.method_ req, path_segments with

  (* GET /api/cards - list with optional ?q= search *)
  | `GET, ["api"; "cards"] ->
    let q = Dream.query req "q" |> Option.value ~default:"" in
    let* rows = Db.collect_list Card_model.get_all_q () in
    (match rows with
     | Error e -> respond_json 500 (`String (Caqti_error.show e))
     | Ok items ->
       let filtered = if q = "" then items
         else List.filter (fun r ->
           let s = (Card_model.row_to_t r).name in String.length s > 0 && (try ignore (Str.search_forward (Str.regexp_string q) s 0); true with Not_found -> false)
           || (match (Card_model.row_to_t r).artist_name with Some s -> String.length s > 0 && (try ignore (Str.search_forward (Str.regexp_string q) s 0); true with Not_found -> false) | None -> false)) items in
       let json = `List (List.map (fun r ->
         let j = Card_model.to_yojson (Card_model.row_to_t r) in
         j) filtered) in
       respond_json 200 json)

  (* POST /api/cards - create *)
  | `POST, ["api"; "cards"] ->
    let* body = Dream.body req in
    (try
       let j = Yojson.Safe.from_string body in
       (match validate_card j with
        | Error errs -> respond_json 422 (`Assoc [("errors", `List (List.map (fun e -> `String e) errs))])
        | Ok () ->
       let params = extract_insert_params j in
       let* ins = Db.exec Card_model.insert_q params in
       (match ins with
        | Error e -> respond_json 500 (`String (Caqti_error.show e))
        | Ok () ->
          let* last_id = Db.find Card_model.last_id_q () in
          (match last_id with
           | Error e -> respond_json 500 (`String (Caqti_error.show e))
           | Ok new_id ->
             let* row = Db.find_opt Card_model.get_by_id_q new_id in
             (match row with
              | Error e -> respond_json 500 (`String (Caqti_error.show e))
              | Ok (Some r) ->
                let j = Card_model.to_yojson (Card_model.row_to_t r) in
                respond_json 201 (j)
              | Ok None -> respond_json 404 (`String "Not found")))))
     with _ -> respond_json 400 (`String "Invalid JSON"))

  (* GET /api/cards/:id - get one *)
  | `GET, ["api"; "cards"; id_str] ->
    (match int_of_string_opt id_str with
     | None -> respond_json 400 (`String "Invalid id")
     | Some id ->
       let* row = Db.find_opt Card_model.get_by_id_q id in
       (match row with
        | Error e -> respond_json 500 (`String (Caqti_error.show e))
        | Ok None -> respond_json 404 (`String "Not found")
        | Ok (Some r) ->
          let j = Card_model.to_yojson (Card_model.row_to_t r) in
          respond_json 200 (j)))

  (* PUT /api/cards/:id - full update *)
  | `PUT, ["api"; "cards"; id_str] ->
    (match int_of_string_opt id_str with
     | None -> respond_json 400 (`String "Invalid id")
     | Some id ->
       let* body = Dream.body req in
       (try
          let j = Yojson.Safe.from_string body in
          (match validate_card j with
           | Error errs -> respond_json 422 (`Assoc [("errors", `List (List.map (fun e -> `String e) errs))])
           | Ok () ->
          let params = extract_insert_params j in
          let (((name, card_type, rarity, mana_cost), (mana_colors, attack, defense, loyalty), (description, flavor_text, image_url, artist_name), (legal_formats, is_banned, is_restricted, power_level)), set_id) = params in
          let upd_params = (((name, card_type, rarity, mana_cost), (mana_colors, attack, defense, loyalty), (description, flavor_text, image_url, artist_name), (legal_formats, is_banned, is_restricted, power_level)), (set_id, id)) in
          let* upd = Db.exec Card_model.update_q upd_params in
          (match upd with
           | Error e -> respond_json 500 (`String (Caqti_error.show e))
           | Ok () ->
             let* row = Db.find_opt Card_model.get_by_id_q id in
             (match row with
              | Error e -> respond_json 500 (`String (Caqti_error.show e))
              | Ok None -> respond_json 404 (`String "Not found")
              | Ok (Some r) ->
                let j = Card_model.to_yojson (Card_model.row_to_t r) in
                respond_json 200 (j))))
       with _ -> respond_json 400 (`String "Invalid JSON")))

  (* POST /api/cards/{id}/ban - behavior ban *)
  | `POST, ["api"; "cards"; id_str; "_id/ban"] ->
    (match int_of_string_opt id_str with
     | None -> respond_json 400 (`String "Invalid id")
     | Some _id ->
       (* TODO: implement behavior ban *)
       respond_json 204 (`Null))

  (* POST /api/cards/{id}/unban - behavior unban *)
  | `POST, ["api"; "cards"; id_str; "_id/unban"] ->
    (match int_of_string_opt id_str with
     | None -> respond_json 400 (`String "Invalid id")
     | Some _id ->
       (* TODO: implement behavior unban *)
       respond_json 204 (`Null))

  (* POST /api/cards/{id}/restrict - behavior restrict *)
  | `POST, ["api"; "cards"; id_str; "_id/restrict"] ->
    (match int_of_string_opt id_str with
     | None -> respond_json 400 (`String "Invalid id")
     | Some _id ->
       (* TODO: implement behavior restrict *)
       respond_json 204 (`Null))

  (* POST /api/cards/{id}/unrestrict - behavior unrestrict *)
  | `POST, ["api"; "cards"; id_str; "_id/unrestrict"] ->
    (match int_of_string_opt id_str with
     | None -> respond_json 400 (`String "Invalid id")
     | Some _id ->
       (* TODO: implement behavior unrestrict *)
       respond_json 204 (`Null))

  (* GET /api/cards/{id}/value - behavior calculate_value *)
  | `GET, ["api"; "cards"; id_str; "_id/value"] ->
    (match int_of_string_opt id_str with
     | None -> respond_json 400 (`String "Invalid id")
     | Some _id ->
       (* TODO: implement behavior calculate_value *)
       respond_json 204 (`Null))

  (* POST /api/cards/{id}/rarity-bonus - behavior apply_rarity_bonus *)
  | `POST, ["api"; "cards"; id_str; "_id/rarity-bonus"] ->
    (match int_of_string_opt id_str with
     | None -> respond_json 400 (`String "Invalid id")
     | Some _id ->
       (* TODO: implement behavior apply_rarity_bonus *)
       respond_json 204 (`Null))

  (* GET /api/cards/{id}/legal - behavior is_legal_in_format *)
  | `GET, ["api"; "cards"; id_str; "_id/legal"] ->
    (match int_of_string_opt id_str with
     | None -> respond_json 400 (`String "Invalid id")
     | Some _id ->
       (* TODO: implement behavior is_legal_in_format *)
       respond_json 204 (`Null))

  | _ -> respond_json 404 (`String "Not found")
